vscode-json-languageserver: migrate from nodePackages

ryota2357 opened the pull request:
Migrate vscode-json-languageserver from nodePackages to pkgs/by-name.

See #229475.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ryota2357 committed:
vscode-json-languageserver: migrate from nodePackages

ryota2357 commented:
I have a question about the binary name for this package.
The official name is `vscode-json-languageserver`, but the hyphenated version `vscode-json-language-server` is commonly used by other tools (e.g., nvim-lspconfig, vscode-language-servers-extracted). Should we add a symlink for better compatibility?

drupol commented on pkgs/by-name/vs/vscode-json-languageserver/package.nix:
```diff
@@ -0,0 +1,39 @@
+  typescript,
+}:
+
+buildNpmPackage rec {
```
```suggestion
buildNpmPackage (finalAttrs: {
```

I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

And then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent examples of implementations: https://github.com/NixOS/nixpkgs/pulls?q=is%3Apr+is%3Aclosed+finalAttrs+pattern+

Feel free to reach out if you need some assistance.

ryota2357 replied:
fixed.
Thank you for your obvious and details explanation of the `finalAttrs`.
I understand the problem of `rec` and why we should use the `finalAttrs` pattern.

drupol commented on pkgs/by-name/vs/vscode-json-languageserver/package.nix:
```diff
@@ -0,0 +1,39 @@
+  src = fetchFromGitHub {
+    owner = "microsoft";
+    repo = "vscode";
+    rev = "1.101.1";
```
```suggestion
    tag = "1.101.1";
```

ryota2357 replied:
fixed.

ryota2357 committed:
vscode-json-languageserver: add `vscode-json-language-server` symblink for compatibility

ryota2357 commented:
@drupol
Thank you for your review! I've fixed.

I want you to ask one more thing.
I've also added a commit for the compatibility-improving symlink I [mentioned earlier](https://github.com/NixOS/nixpkgs/pull/418336#issuecomment-2989949165).

```nix
  postInstall = ''
    ln -s $out/bin/${finalAttrs.pname} $out/bin/vscode-json-language-server
  '';
```

Is it ok or not? If not, I will remove this change.

drupol commented on pkgs/by-name/vs/vscode-json-languageserver/package.nix:
```diff
@@ -0,0 +1,43 @@
+  dontNpmBuild = true;
+
+  postInstall = ''
+    ln -s $out/bin/${finalAttrs.pname} $out/bin/vscode-json-language-server
```
Please don't use the pname attribute, use string literals.

drupol commented on pkgs/by-name/vs/vscode-json-languageserver/package.nix:
```diff
@@ -0,0 +1,43 @@
+    ln -s $out/bin/${finalAttrs.pname} $out/bin/vscode-json-language-server
+  '';
+
+  meta = {
```
Missing meta.mainProgram

ryota2357 replied:
added `mainProgram = "vscode-json-languageserver";`.

ryota2357 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 418336`
Commit: `0af8fecc28c789bed854ad576d2759cd4985766b`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>vscode-json-languageserver</li>
  </ul>
</details>

davidsierradz commented:
Should we remove `vscode-json-language-server` output from `vscode-langservers-extracted`?

```
error: collision between `/nix/store/vbk9blw37mhp7syzkhy9y5jwdczgsfxp-vscode-langservers-extracted-4.10.0/bin/vscode-json-language-server' and `/nix/store/xkjiq37cdp0ds99rrmsid2vnaims851f-vscode-json-languageserver-1.3.4/bin/vscode-json-language-server'
```
